File: tests/appium/pages/premium_page.py
"""
Premium Page Object — Appium (Android)
"""
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from appium.webdriver.common.appiumby import AppiumBy
import config


class PremiumPage:

    def __init__(self, driver):
        self.driver = driver
        self.wait   = WebDriverWait(driver, config.EXPLICIT_WAIT)

    # ── Locators ──────────────────────────────────────────────
    HEADER_TITLE   = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="Upgrade to PRO"]')
    BACK_BTN       = (AppiumBy.XPATH,
        '//android.widget.ImageView[@content-desc="arrow-back"]')
    PRICE_DISPLAY  = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="3.99"]')
    PRICE_PERIOD   = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="/mo"]')
    UPGRADE_BTN    = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="UPGRADE NOW"]')
    VERIFY_BTN     = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="VERIFY PAYMENT"]')
    FEATURE_UNLIMTED = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="Unlimited AI Threat Scans"]')
    FOOTER_TEXT    = (AppiumBy.XPATH,
        '//android.widget.TextView[contains(@text,"Terms of Service")]')
    BILLING_TEXT   = (AppiumBy.XPATH,
        '//android.widget.TextView[@text="Billed monthly. Cancel anytime."]')

    # ── Actions ───────────────────────────────────────────────
    def wait_for_premium_screen(self):
        self.wait.until(EC.presence_of_element_located(self.HEADER_TITLE))
        return self

    def go_back(self):
        self.wait.until(EC.element_to_be_clickable(self.BACK_BTN)).click()
        return self

    def tap_upgrade_now(self):
        self.wait.until(EC.element_to_be_clickable(self.UPGRADE_BTN)).click()
        return self

    def tap_verify_payment(self):
        self.wait.until(EC.element_to_be_clickable(self.VERIFY_BTN)).click()
        return self

    # ── Assertions ────────────────────────────────────────────
    def is_premium_screen_visible(self) -> bool:
        try:
            self.wait.until(EC.presence_of_element_located(self.HEADER_TITLE))
            return True
        except Exception:
            return False

    def get_displayed_price(self) -> str:
        el = self.wait.until(EC.presence_of_element_located(self.PRICE_DISPLAY))
        return el.text

    def is_upgrade_button_visible(self) -> bool:
        try:
            self.driver.find_element(*self.UPGRADE_BTN)
            return True
        except Exception:
            return False

    def are_all_features_listed(self) -> bool:
        features = [
            "Unlimited AI Threat Scans",
            "Unlimited Forensic Chat",
            "Priority Cloud Sync",
            "Zero-Day Protection",
        ]
        for feat in features:
            els = self.driver.find_elements(
                AppiumBy.XPATH, f'//android.widget.TextView[@text="{feat}"]'
            )
            if not els:
                return False
        return True

File: tests/selenium/pages/PremiumPage.js
/**
 * PremiumPage — Page Object Model (Node.js / selenium-webdriver)
 */

'use strict';

const { By, until } = require('selenium-webdriver');
const config = require('../config');

class PremiumPage {
  /** @param {import('selenium-webdriver').WebDriver} driver */
  constructor(driver) {
    this.driver = driver;
  }

  loc = {
    headerTitle: By.xpath('//*[contains(text(),"Upgrade to PRO")]'),
    priceDisplay:By.xpath('//*[contains(text(),"3.99")]'),
    billingText: By.xpath('//*[contains(text(),"Billed monthly")]'),
    upgradeBtn:  By.xpath('//*[contains(text(),"UPGRADE NOW")]'),
    verifyBtn:   By.xpath('//*[contains(text(),"VERIFY PAYMENT")]'),
    footerText:  By.xpath('//*[contains(text(),"Terms of Service")]'),
  };

  FEATURES = [
    'Unlimited AI Threat Scans',
    'Unlimited Forensic Chat',
    'Priority Cloud Sync',
    'Zero-Day Protection',
  ];

  // ── JS Click helper ──
  async jsClick(locator) {
    const el = await this.driver.wait(until.elementLocated(locator), config.EXPLICIT_WAIT);
    await this.driver.executeScript('arguments[0].scrollIntoView({block:"center"}); arguments[0].click();', el);
  }

  // ── Actions ───────────────────────────────────────────────
  async waitForPremium() {
    await this.driver.wait(until.elementLocated(this.loc.headerTitle), config.EXPLICIT_WAIT);
  }

  async goBack() {
    await this.driver.executeScript('window.history.back();');
    await this.driver.wait(
      until.elementLocated(By.xpath('//*[contains(text(),"Cloud Scans")]')),
      config.EXPLICIT_WAIT,
    );
  }

  async clickUpgradeNow() {
    await this.jsClick(this.loc.upgradeBtn);
  }

  // ── Assertions ────────────────────────────────────────────
  async isPremiumVisible() {
    try {
      await this.driver.wait(until.elementLocated(this.loc.headerTitle), config.SHORT_WAIT);
      return true;
    } catch (_) {
      return false;
    }
  }

  async getPriceText() {
    // Retry loop to handle StaleElementReferenceError
    try {
      return await this.driver.wait(async () => {
        try {
          const el = await this.driver.findElement(this.loc.priceDisplay);
          const txt = await el.getText();
          return txt || false;
        } catch (_) {
          return false;
        }
      }, config.EXPLICIT_WAIT);
    } catch (_) {
      return '';
    }
  }

  async isUpgradeBtnVisible() {
    // Retry loop to handle StaleElementReferenceError from React re-renders
    try {
      return await this.driver.wait(async () => {
        try {
          const el = await this.driver.findElement(this.loc.upgradeBtn);
          return await el.isDisplayed();
        } catch (_) {
          return false;
        }
      }, config.SHORT_WAIT);
    } catch (_) {
      return false;
    }
  }

  async areFeaturesListed() {
    for (const feat of this.FEATURES) {
      const els = await this.driver.findElements(
        By.xpath(`//*[contains(text(),"${feat}")]`),
      );
      if (els.length === 0) return false;
    }
    return true;
  }
}

module.exports = PremiumPage;
